### DISCUSSION TITLE
Multiple installations request for advice

### DISCUSSION THREAD

**Discussion by eawmaas** [xenocrat/chyrp-lite#217] [Q&A] [answered]
I have Chyrp Lite in a subdirectory and I have 5 other subdirectories on my site, some with a lot of pages.
For a long time already I would love to have them all interactive, blog-style (so: Chyrp ofcourse).

I have been playing around a bit and I think I can do this quite easily, but I am not sure about that, I do not want to have to change huge amounts of native Chyrp Code.

Say that I have one main installation in the root and several others in subdirectories, I was thinking of making all the installations ONLY FOR THE user-, groups- and Permissions- tables point to the main database. That way users will be logged in for all of my sites and not only one.

Can you please point me in the right direction where to look to make this happen?
Maybe one separate database only for these three user-related tables?

Or will it require huge modifications to the source-code? (I would like to try to build a module for this, that you can enable in every installation  that lets you point to another database, only for the user-part).

**Comment by xenocrat** [comment]
Hello there,

I think trying to use multiple databases is going to be impossible because of Chyrp Lite's architecture, but there is another option that might work.

Chyrp Lite has the concept of a "table prefix" - this is a unique identifier you can specify during installation, which will be added to all table names created by Chyrp Lite. For example, if you specify the table prefix "myblog_" then Chyrp Lite will create tables named "myblog_posts" instead of "posts", "myblog_pages" instead of "pages", "myblog_users" instead of "users" and so on. This feature allows you to run several blogs using a single database and unique table prefixes.

Now, it _might_ work to modify one method in one file of the core, to make it do something different with this prefixing feature depending on the table being requested. You might be able to make prefixing work as normal, unless the table is one of the ones you want to be shared across installations.

To do this, modify [this line](https://github.com/xenocrat/chyrp-lite/blob/master/includes/class/SQL.php#L178) in _includes/class/SQL.php_ and replace it with:

````
$query = preg_replace_callback("/(__)([a-zA-Z0-9_]+)>/", function ($match) {
    switch ($match[2]) {
        case "users":
        case "groups":
        case "permissions":
        case "sessions":
            return $match[2];
        default:
            return SQL::current()->prefix.$match[2];
    }
}, $query);
````

Now you can run multiple blogs in subdirectories, all using the same database and having their own prefixed tables for most things but sharing their users, groups, permissions, and sessions tables. This will mean that a user logged into one blog is logged into them all, so long as they are served from the same web domain (example.com/blog1/, example.com/blog2/).

I must stress that Chyrp Lite isn't designed to do anything like this, so you will be in uncharted territory. I haven't thought through all the implications of this change so it might just fail for a reason I haven't considered yet. But I will try to help as much as possible. Good luck!

**Reply by eawmaas** [reply]
Thank you! I am going to dive into it, don't expect results tomorrow morning, it might take some time but I am progressing still and feel confident and determined about this, what you proposed (if I can get it working) would be a solution for many projects in the future for me because Chyrp has it all, why invent the wheel twice? 👍

**Comment by eawmaas** [comment]
Hi!

I have been working on this and experimenting and think it can be done but because Chyrp is not designed to do this I think it wouldn't be the best option for the future. New releases could break it or cause trouble.

Originally I wanted all my websites to be interactive but I couldn't find a good and secure way until I stumbled upon Chyrp Lite, now I am thinking of starting again from scratch and rebuilding everything. For this I would love to use the feature that you can make pages and subpages etc. but I really need an option to **enable comments on pages**.
Optionally a module that enables **PHP-scripting in pages**

I have a feeling that this could be done pretty easily, am I right?

**Reply by xenocrat** [reply]
Hello there,

Enabling comments on pages is actually more difficult than it seems, because comment objects are designed with the expectation that they belong to a post. I see the usefulness of having this feature so I'll give some thought to how it could be done.

I don't want to build in support for PHP scripting in pages (or any other Twig template files), because it gives theme authors the ability to accidentally cause far more numerous and more serious security vulnerabilities with simple coding mistakes. However if you known what you are getting into, you could create a module that defines a new Twig function to `eval()` some PHP supplied to it. But what PHP code would you want to add to the page? There is probably a better way to achieve your desired result.

**Reply by eawmaas** [reply]
Well, it would be very nice if I can query another database that contains genealogy-data, so that I can build a family-tree application inside Chyrp. I have all the data in a MySQL database, say that I make a page with different functions and include that page then I could use the functions in a blogpost or blogpage. These functions will mainly be MySQL-queries and building the reports. I will use css to display the results in the blogpost.

Thing is: there are different applications that you can install to display a familytree, but none of them is interactive like in a blog and since Chyrp Lite has all these facilities ánd a user-group system I thought it would be nice if I can like hook functionalities from php-files outside Chyrp into a blogpost or page, that's why I would like to have the possibility to add comments to pages as well because I was thinking of making pages in hierarchical structures for this, which fits genealogy better.

So, simply thought: include a page called something like genealogyFunctions.php and retrieve results from the genealogy-database.

Maybe even in the future add a group to Chyrp that gives family and friends who register the right to view persons who are still alive (=privacy-regulations).

Chyrp has all the functionalities to become a genealogy-blog, I just need to find a way to add some kind of communication with the genealogy-database, using external functions would be the easiest way I guess, that can be procedural PHP

**Reply by xenocrat** [reply, marked as answer]
I think this would be a good use case for additional Twig functions defined by a module. Here's how it works: You create a [basic module](https://github.com/xenocrat/chyrp-lite/wiki/Making-Your-First-Module) that has a few functions named like `twig_function_<<your Twig function name>>`, for example `twig_function_render_family_tree()`.

Then in any Twig template, you can call the function `{{ render_family_tree() }}` to do some work and output any HTML you desire, or a filter `{{ start_position | render_family_tree }}` if you prefer. You can see exactly how this works by inspecting [Leaf](https://github.com/xenocrat/chyrp-lite/blob/master/includes/lib/Leaf.php), Chyrp Lite's extension class. Functions and filters defined by a module would work just like the built-in functions `mailto()` and `paginate()`, or the filters `thumbnail` and `translate` - they can query any external resources you like, they can receive parameters when called from the Twig template, e.g. `render_tree($start_variable, $depth_variable)`, and they have full awareness of the currently logged in user and the group that user belongs to (via the __Visitor__ class) so they can self-censor in a similar way to posts and pages, showing only what is appropriate for the currently logged-in user.

If you look through the file _Leaf.php_, you'll get a good sense for the range of possibilities that custom Twig functions and filters make possible. I think you could achieve everything you describe using these capabilities.

**Reply by eawmaas** [reply]
Thank you! This sounds great and a bit as I thought of as an alternative way and maybe even the appropriate way yes, you confirm this with your answer.

I have been doing some extensive reading into twig and basically it has all functionalities like loops, variables, functions etc. yes.

I am looking forward to explore this route of creating what I have in mind! I am not in a hurry, at the moment I am switching from Windows to Linux which doesn't go as smooth as I expected.

But I know now exactly what to do.
